src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -265,6 +265,23 @@ public ValueTask<FlushResult> WriteDataAsync(int streamId, StreamOutputFlowContr
             }
         }
 
+        public void WriteDataWithoutFlush(int streamId, StreamOutputFlowControl flowControl, in ReadOnlySequence<byte> data, bool endStream)
```

Comment:
Comment that flow control availability must be checked by the caller.

Reply:
`dataLength > flowControl.Available` is being checked by the caller outside of `lock (_writeLock)`. Problem?

Reply:
I think it is an issue. flowControl is only accessed when holding the _writeLock, seems very prone to an issue.

Reply:
Hmm, probably. Your writes for this stream are ordered and avoid stream level flow control issues, but you could have competing writes from other streams and mess up the connection level flow control.

Reply:
Maybe add a parent method that checks Available under the lock?
- If available then call and call Advance still under the lock and then call WriteDataWithoutFlush
- If not then call WriteDataAsync as needed

Reply:
I've removed the new method, and made forcing flushing an argument on the existing method. The logic between the two paths is exactly the same apart from calling `TimeFlushUnsynchronizedAsync`.